src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs

```
@@ -37,49 +38,78 @@ public Task<T> AddInvocation<T>(string connectionId, string invocationId, string
+            var result = _pendingInvocations.TryAdd(invocationId, 
+                new PendingInvocation(
+                    typeof(T), connectionId, instanceId, tcs, 
+                    static (state, completionMessage) =>
+                    {
+                        var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+                        if (completionMessage.HasResult)
+                        {
+                            tcs.SetResult((T)completionMessage.Result);
+                        }
+                        else
+                        {
+                            tcs.SetException(new Exception(completionMessage.Error));
+                        }
+                    })
+            );
+            Debug.Assert(result);
+
+            tcs.RegisterCancellation();
+
+            return tcs.Task;
+        }
+
+        public void AddServiceMapping(ServiceMappingMessage serviceMappingMessage)
+        {
+            if (_pendingInvocations.TryGetValue(serviceMappingMessage.InvocationId, out var invocation))
             {
-                var tcs = (TaskCompletionSourceWithCancellation<T>)state;
-                if (completionMessage.HasResult)
+                if (invocation.RouterInstanceId == null)
                 {
-                    tcs.SetResult((T)completionMessage.Result);
+                    invocation.RouterInstanceId = serviceMappingMessage.InstanceId;
+                    _pendingInvocations[serviceMappingMessage.InvocationId] = invocation;
                 }
                 else
                 {
-                    tcs.SetException(new Exception(completionMessage.Error));
+                    throw new InvalidOperationException($"Failed to record a service mapping whose RouterInstanceId '{serviceMappingMessage.InvocationId}' was already existing.");
```

silent failure is enough?